### PR TITLE
validate_build over settings

### PR DESCRIPTION
**Pull request by lasote**
Changelog: Feature: Introduced a `validate_build()` method in the recipes to validate if the package can be built according to `self.settings` and `self.options` instead of `self.info.settings` and `self.info.options` that it is used to validate the `binary`.
Docs: https://github.com/conan-io/docs/pull/2667

### CODE REVIEW

File: conans/client/graph/graph_binaries.py [hunk 1]
```diff
@@ -451,3 +455,14 @@ def reevaluate_node(self, node, remotes, build_mode, update):
...
+
+
+def _call_validate_build(node):
+    conanfile = node.conanfile
+    if hasattr(conanfile, "validate_build") and callable(conanfile.validate_build):
+        with conanfile_exception_formatter(str(conanfile), "validate_build"):
+            try:
+                conanfile.validate_build()
+            except ConanInvalidConfiguration as e:
+                conanfile.info.invalid = str(e)
```
Comment (memsharded) [inline conans/client/graph/graph_binaries.py]: This is weird. If it cannot be built, it doesn't depend on the ``info`` for ``package_id``. It is a property of the ``conanfile.cant_build = str(e)``. Overriding ``info.invalid`` can destroy information of the potential valid/invalid of the ``package_id``

File: conans/client/graph/graph_binaries.py [hunk 2]
```diff
@@ -295,6 +298,7 @@ def _process_node(self, node, pref, build_mode, update, remotes):
                     conanfile.output.info("Outdated package!")
                     node.binary = BINARY_BUILD
                     node.prev = None
+                    _call_validate_build(conanfile)
```
Comment (memsharded) [inline conans/client/graph/graph_binaries.py]: I would try to call this earlier in the process, just once, as it doesn't have much effect besides setting an attribute, it should be possible.

**Review by memsharded — COMMENTED** [review]
Too many tests are broken, but I fail to see how these changes can cause that breakage

File: conans/client/graph/graph_binaries.py (hunk 1 above)
Comment (lasote) [reply conans/client/graph/graph_binaries.py]: ok, this is a copy paste I didn't know anything about the `cant_build`

File: conans/client/graph/graph_builder.py [hunk 3]
```diff
@@ -471,4 +471,12 @@ def _create_new_node(self, current_node, dep_graph, requirement, check_updates,
 
         dep_graph.add_node(new_node)
         dep_graph.add_edge(current_node, new_node, requirement)
+
+        if hasattr(dep_conanfile, "validate_build") and callable(dep_conanfile.validate_build):
```
Comment (lasote) [inline conans/client/graph/graph_builder.py]: Probably do the `validate_build()` here is too soon to have the final options.

Comment (memsharded) [reply conans/client/graph/graph_builder.py]: It is, it will not have the final options values.

File: conans/client/graph/graph_manager.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
 from conans.client.graph.graph_binaries import RECIPE_CONSUMER, RECIPE_VIRTUAL, BINARY_EDITABLE, \
     BINARY_UNKNOWN
 from conans.client.graph.graph_builder import DepsGraphBuilder
-from conans.errors import ConanException, conanfile_exception_formatter
+from conans.errors import ConanException, conanfile_exception_formatter, ConanInvalidConfiguration
```
Comment (memsharded) [inline conans/client/graph/graph_manager.py]: unused?

File: conans/client/graph/graph_builder.py (hunk 3 above)
Comment (lasote) [reply conans/client/graph/graph_builder.py]: Moved, please review. I don't really know the best place to do it.

**Review by memsharded — APPROVED** [review]
Looks good.

Don't forget to mark it as experimental in the docs

**Review by czoido — APPROVED** [review]

**Pull request merged**